### PR TITLE
Add a function that writes to a single HDF5 file in parallel across multiple locales

### PR DESCRIPTION
**Pull request by daviditen**
This adds a function for writing a block distributed array to a single HDF5
file in parallel across multiple locales. The data can later be read back
with a different number of locales.  It uses HDF5 collective operations
which use MPIIO under the covers.

This function currently only handles block distributed arrays, while the
equivalent reading function handles both block and cyclic. I think this can
be updated to also handle cyclic by giving the file dataset view non-unit
stride.

There are some extern declarations inside the `hdf5WriteDistributedArray`
function that should likely be in the C_MPI or C_HDF5 modules, but I left
them here for now.

This requires the hdf5-parallel library as well as MPI.

### CODE REVIEW

**Comment by ronawho** [comment]
There can be a non-trivial amount of contention/overhead from having an MPI runtime co-exist with ours. Could you try running stream on an XC with and without writing the `A` array to a file before doing the main kernel?

**Comment by daviditen** [comment]
I tried modifying stream.chpl to add HDF5 to the list of used modules and added this just after printConfiguration():
```
  {
    var writeme = newBlockArr({1..100, 1..100}, int);
    hdf5WriteDistributedArray("mydata.h5", "dset", writeme);
  }
```

Here's with a writing a 100x100 array to an HDF5 file before the timed part:
```
% ./stream-write -nl16 --m=5723827200
Number of Locales = 16
Tasks per locale = 0
Problem size = 5723827200
Bytes per array = 45790617600
Total memory required (GB) = 127.938
Number of trials = 10

Validation: SUCCESS
Execution time:
  tot = 2.81482
  avg = 0.281482
  min = 0.279895
Performance (GB/s) = 490.798
```

Here's unmodified stream.chpl:
```
% ./stream-no-write -nl16 --m=5723827200
Number of Locales = 16
Tasks per locale = 0
Problem size = 5723827200
Bytes per array = 45790617600
Total memory required (GB) = 127.938
Number of trials = 10

Validation: SUCCESS
Execution time:
  tot = 3.0228
  avg = 0.30228
  min = 0.296352
Performance (GB/s) = 463.543
```

**Comment by ronawho** [comment]
I ran a similar variation that wrote A before and after. Stream diff:

<details>

```diff
diff --git a/test/release/examples/benchmarks/hpcc/stream.chpl b/test/release/examples/benchmarks/hpcc/stream.chpl
index 0430fc3894..be98bacfc5 100644
--- a/test/release/examples/benchmarks/hpcc/stream.chpl
+++ b/test/release/examples/benchmarks/hpcc/stream.chpl
@@ -71,6 +71,8 @@ proc main() {

   var execTime: [1..numTrials] real;                 // an array of timings

+  use HDF5;
+  hdf5WriteDistributedArray("before.h5", "dset", A);
   for trial in 1..numTrials {                        // loop over the trials
     const startTime = getCurrentTime();              // capture the start time

@@ -84,6 +86,7 @@ proc main() {

     execTime(trial) = getCurrentTime() - startTime;  // store the elapsed time
   }
+  hdf5WriteDistributedArray("after.h5", "dset", A);

   const validAnswer = verifyResults(A, B, C);        // verify...
   printResults(validAnswer, execTime);               // ...and print the results
```

</details>

I saw no performance difference on 16 nodes of an XC. And that run competed in under 5 minutes, which is pretty awesome considering it was writing (2) 185 GB arrays

File: modules/packages/HDF5.chpl [hunk 1]
```diff
@@ -3784,30 +3784,34 @@ module HDF5 {
...
-  }
-
   /* Write the Block distributed Array `A` as an HDF5 dataset named `dsetName`
      in the file `filename` using parallel collective IO. This requires the
-     hdf5-parallel library.
+     hdf5-parallel library, which requires the MPI library.
+
+     The file written by this function can be read in parallel with the
+     function `hdf5ReadDistributedArray`. The write and read operations
+      can use arrays distributed over different numbers of locales.
```
Comment (vasslitvinov) [inline modules/packages/HDF5.chpl]: Minor: an extra leading space. Might confuse RST (or not).

**Comment by bradcray** [comment]
@daviditen: It appears that this change requires linking MPI into Chapel for any use of the HDF5 module, even if the program does not call the specific routines that require MPI.  For example, building Arkouda's master branch with Chapel's master branch results in MPI link errors even though (I don't think?) Arkouda uses routines that require MPI.  This makes me wonder whether we might want to tease the pieces that do require MPI into a distinct module or otherwise only require MPI when the routines in question are called.  You're much more familiar with this than I am, what do you think about this?

**Comment by daviditen** [comment]
```chapel
proc uncalled() {
  use MPI;
}

proc main { }
```

To me it feels like a bug that this code would call the initializer for the MPI module. Something like this is why using the HDF5 module now requires linking with MPI even if you don't actually use anything in HDF5. The `uncalled` function is completely removed after resolution, but still causes the MPI module to be initialized.  Fixing this would make linking MPI only be required if the function that uses it is actually called.

**Comment by bradcray** [comment]
That's a good point.  Would you be up for creating an issue capturing this idea and adding an exploration of making that improvement to your TODO list?  It seems like it could also help with compile times (at least slightly) by reducing the amount of code generated for modules that aren't actually required by the final executable (?).  Tagging @ben-albrecht on this because it feels vaguely related to some discussions we had awhile back regarding numerical libraries and their dependencies on various libraries.

**Comment by ben-albrecht** [comment]
Yes, here's the current work-around implemented for the LinearAlgebra/BLAS/LAPACK interfaces:

config params determine which BLAS implementation is used, if any at all: 

- https://github.com/chapel-lang/chapel/blob/master/modules/packages/BLAS.chpl#L177-L209

params in LinearAlgebra depends on this and are used as where clauses on functions that use BLAS calls: 

- https://github.com/chapel-lang/chapel/blob/master/modules/packages/LinearAlgebra.chpl#L192-L194
- https://github.com/chapel-lang/chapel/blob/master/modules/packages/LinearAlgebra.chpl#L646-L647

There was some discussion about other potential solutions in https://github.com/chapel-lang/chapel/issues/8803

**Comment by daviditen** [comment]
I opened issue #14503.

File: modules/packages/HDF5.chpl [hunk 2]
```diff
@@ -3784,11 +3784,134 @@ module HDF5 {
...
 
+  /* Write the Block distributed Array `A` as an HDF5 dataset named `dsetName`
+     in the file `filename` using parallel collective IO. This requires the
+     hdf5-parallel library, which requires the MPI library.
+
+     The file written by this function can be read in parallel with the
+     function `hdf5ReadDistributedArray`. The write and read operations
+     can use arrays distributed over different numbers of locales.
+   */
+  proc hdf5WriteDistributedArray(A: [], filename: string, dsetName: string) {
```
Comment (bradcray) [inline modules/packages/HDF5.chpl]: In today's demo, I found myself thinking "What's the point of using an hdf5 prefix on this function given that it's defined in the HDF5 module?"  It seems like a very flat, C-based naming scheme for a language that doesn't need it.

**Pull request merged**